memory usage very high and not exit when code has some compile error

### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/usage/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

go install

### Description of the problem

We found the lint stage take many memory and got killed in the CI recently.

<img width="635" alt="Image" src="https://github.com/user-attachments/assets/243a5cea-85f1-4070-8a66-1bafb93ca521" />

### Version of golangci-lint

<details>

```console
golangci-lint has version  built with go1.24.0 from (unknown, modified: ?, mod sum: "") on (unknown)
```

</details>

### Configuration

<details>

```console
../golangci-lint/golangci-lint-debug run ./...
```

</details>

### Go environment

<details>

```console
go version go1.24.0 darwin/amd64
AR='ar'
CC='clang'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='clang++'
GCCGO='gccgo'
GO111MODULE=''
GOAMD64='v1'
GOARCH='amd64'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/xxx/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/xxx/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/bh/g2h9d2vx35b2b4k8pn4k2rt80000gn/T/go-build664656937=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='amd64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/Users/xxx/work/xxx/yyy/kubernetes/go.mod'
GOMODCACHE='/Users/xxx/go/pkg/mod'
GONOPROXY='git.in.xxx.com'
GONOSUMDB='git.in.xxx.com'
GOOS='darwin'
GOPATH='/Users/xxx/go'
GOPRIVATE='git.in.xxx.com'
GOPROXY='https://goproxy.in.xxx.com,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTELEMETRY='on'
GOTELEMETRYDIR='/Users/xxx/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/darwin_amd64'
GOVCS=''
GOVERSION='go1.24.0'
GOWORK='/Users/xxx/work/xxx/yyy/kubernetes/go.work'
PKG_CONFIG='pkg-config'
```

</details>

### Verbose output of running

<details>

```console
```

</details>

### A minimal reproducible example or link to a public repository

I have reproduced the bug in  https://github.com/alingse/golangci-lint/tree/dump-heap  and  https://github.com/alingse/kubernetes/tree/add-small-crack , but I don't know is this only way to trigger the bug.

only a small compile error in "_test.go"  cause the golangci-lint take 25GB memory and keep growing


<details>

<img width="554" alt="Image" src="https://github.com/user-attachments/assets/07048be7-d694-47b6-be9e-2433dc611266" />

<img width="841" alt="Image" src="https://github.com/user-attachments/assets/0defe386-4ed7-4821-a4da-5675657cbd96" />

<img width="281" alt="Image" src="https://github.com/user-attachments/assets/58e1394a-cc10-490e-9156-5647fc7f33f0" />

<img width="1440" alt="Image" src="https://github.com/user-attachments/assets/c0a56293-1c18-4380-9950-8905fbd60c9b" />

</details>

I research the code and pprof, found this code will create a tree of `joinError`, and cause memory use when `fmt.Errorf`

the check has alread done, but the string format process never exit (maybe I wait long time it can return)

```go
	var depErrors error
	for _, dep := range act.Deps {
		if dep.Err != nil {
			depErrors = errors.Join(depErrors, errors.Unwrap(dep.Err))
		}
	}
	if depErrors != nil {
		act.Err = fmt.Errorf("failed prerequisites: %w", depErrors)
		return
	}
```

the `errror.Join` not implement the `func Unwrap() error`, it only implement the `func Unwrap() []error`

> // In particular Unwrap does not unwrap errors returned by [Join].

and `fmt.Errorf("failed prerequisites: %w", depErrors)` will format every error `Error()` string in the tree.


I try change the code like this ( https://github.com/golangci/golangci-lint/commit/b7bc902fcfb8c0f3d451adf5f931abdd50ea1b99 ) , it exit and report compile error now.

```go
	var depErrors []error
	for _, dep := range act.Deps {
		if dep.Err != nil {
			depErrors = append(depErrors, errors.Unwrap(dep.Err))
		}
	}
	if depErrors != nil {
		act.Err = errors.Join(depErrors...)
		return
	}
```

```console
test/integration/cronjob/cronjob_test.go:19:1: expected 'package', found 'import' (typecheck)
import (
^
test/integration/cronjob/cronjob_test.go:20:2: expected ';', found "context" (typecheck)
	"context"
	^
2 issues:
* typecheck: 2
```



### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

--- a/pkg/goanalysis/runner_checker.go
+++ b/pkg/goanalysis/runner_checker.go
@@ -90,12 +90,12 @@ func (act *action) analyze() {
 	// Report an error if any dependency failures.
 	var depErrors error
 	for _, dep := range act.Deps {
-		if dep.Err != nil {
+		if dep.Err != nil && errors.Unwrap(dep.Err) != nil {
 			depErrors = errors.Join(depErrors, errors.Unwrap(dep.Err))
 		}
 	}
 	if depErrors != nil {
-		act.Err = fmt.Errorf("failed prerequisites: %w", depErrors)
+		act.Err = errors.Join(depErrors, errors.New("failed prerequisites"))
 		return
 	}
 

